Slice in lab frame for boosted sim

RevathiJambunathan opened the pull request:
Implemented slice diagnostic to write out the back-transformed field and particle data for analysis in lab-frame. 

The input file for the slice is added as follows similar to the 3D boosted frame diags.
```
slice.dom_lo = -30.0e-6  0.0 -56e-6 
slice.dom_hi = 30.0e-6   0.0  12e-6
slice.coarsening_ratio = 1 1 1
slice.plot_int = 1000
slice.num_slice_snapshots_lab = 7
slice.dt_slice_snapshots_lab = 1.6678204759907604e-12 # Every 2 cm - as in 2 stages run
```
The slice output plotfiles are in `lab_frame_data/slices`
and the 3D snapshots are in `lab_frame_data/snapshots`.

RevathiJambunathan commented on Source/WarpX.cpp:
```diff
@@ -65,6 +65,9 @@ Real WarpX::dt_snapshots_lab  = std::numeric_limits<Real>::lowest();
 bool WarpX::do_boosted_frame_particles = true;
 
+int  WarpX::num_slice_snapshots_lab = 0;
+Real WarpX::dt_slice_snapshots_lab  = std::numeric_limits<Real>::lowest();
```
Thank you for this suggestion. I have made the change.

RevathiJambunathan commented on Source/WarpX.cpp:
```diff
@@ -543,6 +546,13 @@ WarpX::ReadParameters ()
+          AMREX_ALWAYS_ASSERT_WITH_MESSAGE(gamma_boost > 1.0, 
+                 "gamma_boost must be > 1 to use the boost frame diagnostic");
+          pp.query("num_slice_snapshots_lab", num_slice_snapshots_lab);
+          pp.query("dt_slice_snapshots_lab", dt_slice_snapshots_lab );          
```
Done.

RevathiJambunathan commented on Source/Diagnostics/WarpXIO.cpp:
```diff
@@ -408,7 +408,8 @@ WarpX::InitFromCheckpoint ()
 
-    BL_PROFILE("WarpX::GetCellCenteredData");
+    BL_PROFILE_REGION("GetCellCenteredData()");
+//    BL_PROFILE("WarpX::GetCellCenteredData");
```
Yes, Done.

RevathiJambunathan commented on Source/Diagnostics/BoostedFrameDiagnostic.H:
```diff
@@ -20,55 +20,106 @@
-    };
+class LabFrameDiag { 
+   public:
+   std::string file_name;
```
Fixed it to have 4 spaces.

MaxThevenet commented:
Thanks for this PR!
Could you fix the failing end-of-line whitespaces test? If you click on `Details` on the line 
```
source / style (pull_request) Failing after 11s
```
above, @ax3l was kind enough to suggest some commands you can execute to fix the issue.

RevathiJambunathan commented:
Yes absolutely. :)

MaxThevenet commented on Source/Diagnostics/BoostedFrameDiagnostic.H:
```diff
@@ -109,9 +167,10 @@ public:
-                           const amrex::Geometry& geom);
+                           const amrex::Geometry& geom, amrex::RealBox& slice_realbox);
 
     void Flush(const amrex::Geometry& geom);
```
Could you add a quick description (1 line is often enough) to all these member functions? For instance, a reader might not understand why there is both a `Flush` and a `write` function, and the reason is that
 - `writeLabFrameData` fills the buffers and writes to disk when buffers are full
 - `Flush` is only called at the end of the simulation, and writes the buffers even if they're not full.

MaxThevenet replied:
Actually `writeLabFrameData` is doing a lot of things, could you briefly describe it (either in the `cpp` of `H` file)? I know this is not really linked to your PR, but it's always good to add helpful comments :)

RevathiJambunathan replied:
I have added a brief description in the .H file. 
Please let me know if you have any additions/suggestions

MaxThevenet commented on Source/Diagnostics/BoostedFrameDiagnostic.cpp:
```diff
@@ -690,97 +766,130 @@ writeLabFrameData(const MultiFab* cell_centered_data,
+            //                 LabFrameDiags_[i+1]->t_lab != prev_t_lab ) {
+            //   tmp_slice_ptr.reset(new MultiFab());
+            //   tmp_slice_ptr.reset(nullptr);
+            //}
```
Can we get rid of the commented lines above?

RevathiJambunathan replied:
Done. Thanks for catching this.

MaxThevenet commented on Source/Diagnostics/BoostedFrameDiagnostic.cpp:
```diff
@@ -589,15 +601,79 @@ BoostedFrameDiagnostic(Real zmin_lab, Real zmax_lab, Real v_window_lab,
+    IntVect slice_ncells_lab ;
+    Box slicediag_box;
+
+    if ( N_slice_snapshots_ > 0) {
```
Could you let us know what you are doing here, if ( N_slice_snapshots_ > 0)?

RevathiJambunathan replied:
I have added some description above the code line describing that if reduced diagnostics is used then we obtain the Box for the diag using the lo and hi and the corresponding indices of the lab-frame 'slice' (Which can be 1D, 2D, or 3D).

MaxThevenet commented on Source/Diagnostics/BoostedFrameDiagnostic.H:
```diff
@@ -20,55 +20,123 @@
+    ///
+    void writeLabFrameHeader();
+
+    virtual int getId() {return 0;}
```
There are a number of member functions and variables that are not used. For instance, I could not find any call to `getId`, and any place where `snapshot_id` and `slice_id` are used. Did I miss something? If not, could you do some cleaning and remove *all* the unused members? Compiling with `DEBUG=TRUE` might help, but I am not sure it catches everything.

RevathiJambunathan replied:
Thank you for this suggestion. I have now removed unused variables and functions detected using the DEBUG flag.

MaxThevenet commented on Source/Diagnostics/BoostedFrameDiagnostic.cpp:
```diff
@@ -917,13 +1025,15 @@ writeMetaData ()
-
-        if (!UtilCreateDirectory(WarpX::lab_data_directory, 0755))
-            CreateDirectoryFailed(WarpX::lab_data_directory);
+        //const std::string fullpath = WarpX::lab_data_directory + "/snapshots";
```
Could we get rid of this line?

RevathiJambunathan replied:
Done. replaced  
const std::string fullpath = WarpX::lab_data_directory ; with 
const std::string fullpath = WarpX::lab_data_directory + "/snapshots";

MaxThevenet commented on Source/Diagnostics/BoostedFrameDiagnostic.cpp:
```diff
@@ -917,13 +1025,15 @@ writeMetaData ()
         VisMF::IO_Buffer io_buffer(VisMF::IO_Buffer_Size);
         std::ofstream HeaderFile;
         HeaderFile.rdbuf()->pubsetbuf(io_buffer.dataPtr(), io_buffer.size());
+        //std::string HeaderFileName(WarpX::lab_data_directory + "/snapshots/Header");
```
Could we get rid of this line?

RevathiJambunathan replied:
The new location for the back-transformed lab-frame diagnostic of the full domain is in lab_fame_data/snapshots, so the header file location is written in the same directory. 
 std::string HeaderFileName(WarpX::lab_data_directory + "/snapshots/Header");
I have removed the previous path ( std::string HeaderFileName(WarpX::lab_data_directory + "/Header");)

MaxThevenet commented on Source/Diagnostics/BoostedFrameDiagnostic.cpp:
```diff
@@ -937,31 +1047,78 @@ writeMetaData ()
+   Real zmin_lab = prob_domain_lab_.lo(AMREX_SPACEDIM-1);
+   initial_i = (current_z_lab - zmin_lab) / dz_lab_ ;
+   //file_name = Concatenate(WarpX::lab_data_directory + "/snapshots/snapshot",
+   //                        file_num, 5);
```
Could we get rid of these lines?

RevathiJambunathan replied:
I have kept  
  file_name = Concatenate(WarpX::lab_data_directory + "/snapshots/snapshot",
                           file_num, 5);
and removed 
   file_name = Concatenate(WarpX::lab_data_directory + "/snapshot",
                           file_num, 5);

RevathiJambunathan commented on Source/Diagnostics/WarpXIO.cpp:
```diff
@@ -407,7 +407,7 @@ WarpX::InitFromCheckpoint ()
 WarpX::GetCellCenteredData() {
 
-    BL_PROFILE("WarpX::GetCellCenteredData");
+    BL_PROFILE_REGION("GetCellCenteredData()");
```
I added BL_PROFILE_REGION when I was timing GetCellCenteredData() to see how much time it takes and if using a slice to obtain the CellCenteredData() will help reduce the time. Because as of now -- the steps performed for the full 3D snapshot and slice is the same. It was only for my own analysis -- I will revert it back.

RevathiJambunathan commented on Source/Diagnostics/BoostedFrameDiagnostic.cpp:
```diff
@@ -690,97 +766,130 @@ writeLabFrameData(const MultiFab* cell_centered_data,
-            tmp.copy(*slice, 0, 0, ncomp);
-
-            // Copy data from MultiFab tmp to MultiDab data_buffer[i]
-            CopySlice(tmp, *data_buffer_[i], i_lab, map_actual_fields_to_dump);
```
Oh yes. Thank you

RevathiJambunathan commented on Source/Diagnostics/BoostedFrameDiagnostic.H:
```diff
@@ -20,55 +20,123 @@
+
+    virtual int getId() {return 0;}
+    virtual void AddDataToBuffer(amrex::MultiFab& tmp_slice_ptr, int i_lab,
+                 amrex::Gpu::ManagedDeviceVector<int> map_actual_fields_to_dump){};
```
Yes, this is true for the full domain diagnostic. 
For the reduced diagnostic, the data is selectively copied if the extent of the z_lab temporary multifab intersects with the user-defined sub-domain for the reduced diagnostic (i.e., a 1D, 2D, or 3D region of the domain).

RevathiJambunathan commented on Source/Diagnostics/BoostedFrameDiagnostic.H:
```diff
@@ -20,55 +20,123 @@
+    void AddPartDataToParticleBuffer( amrex::Vector<WarpXParticleContainer::DiagnosticParticleData> tmp_particle_buffer, int nSpeciesBoostedFrame);
+};
+
+class BoostedFrameDiagnostic {
```
I have added a brief description of this class.

RevathiJambunathan commented on Source/Diagnostics/BoostedFrameDiagnostic.H:
```diff
@@ -80,23 +148,13 @@ class BoostedFrameDiagnostic {
-    amrex::Vector<int> buff_counter_;
 
-    amrex::Vector<LabSnapShot> snapshots_;
+    std::vector<std::unique_ptr<LabFrameDiag> > LabFrameDiags_;
```
Yes. LabFrameDiags contains all snapshots and slices. I have now added a description of this class and the derived classes.

RevathiJambunathan commented on Source/WarpX.cpp:
```diff
@@ -588,6 +591,15 @@ WarpX::ReadParameters ()
+          pp.query("num_slice_snapshots_lab", num_slice_snapshots_lab);
+          if (num_slice_snapshots_lab > 0) {
+             pp.get("dt_slice_snapshots_lab", dt_slice_snapshots_lab );
+          }
```
Yes, the code runs with the default value of 0 for num_slice_snapshots_lab

RevathiJambunathan commented:
> Thank you for this PR. Could you please address the points below and the comments in the PR?
> 
>     1. Could you split long lines (>80 characters), when possible? For instance,
> 
> 
> ```c++
>     void AddPartDataToParticleBuffer( amrex::Vector<WarpXParticleContainer::DiagnosticParticleData> tmp_particle_buffer, int nSpeciesBoostedFrame);
> ```
> 
> could be
> 
> ```
>     void AddPartDataToParticleBuffer(
>         amrex::Vector<WarpXParticleContainer::DiagnosticParticleData> tmp_particle_buffer,
>         int nSpeciesBoostedFrame);
> ```
> 
>     1. Could you put the constructor declaration at the top of the declaration of each class? I think that's pretty much what is done in the rest of the code, and it seems standard.
> 
>     2. Could you add the documentation on how to use this feature, and describe the options that are (or are not) available? For instance, can we create a 2D slice for a 3D simulation, and read the 2D slice in yt? (Sorry I know I asked this question some time ago, but I don't remember the answer...).
> 
>     3. Could we write an automated test for this feature? For instance, we could run a 3D simulation with back-transformed diagnostics and dump both 3D data and some slices. Then we could read the 3D data, take the central slice in post-processing and compare it with the site diagnostics.
> 
>     4. There seems to be some comment old code. Could you clean them?
> 
>     5. Did you check that these changes are compatible with the HDF5 implementation?

RevathiJambunathan commented:
Hi Maxence --

Thank you for your comments and suggestions. 
1. I have now tried to ensure that characters per line <=80 (more or less)
    
2. For the new classes, I have put the constructor declaration at the top of the declaration of each class. 

3. Included documentation on slice parameters and mentioned in visualization on how to plot the back-transformed field data for the full diagnostic as well as reduced 'slice' diagnostic. 
Please let me know if more information is required. 


4. I have added an automated test with input file in Examples/Modules/RigidInjection/input.BoostedFrame_testSlice and the corresponding error check in  Examples/Modules/RigidInjection/analysis_3Dbacktransformed_diag.py
The path to these files are added under [[Boost_3Dbacktransformed_CheckReducedDiagnostics] in Regression/WarpX-tests.ini

5. I have now cleaned the code.. and the snapshots and slices are written in lab_frame_data/snapshots and lab_frame_data/slices, respectively. Please see PR#383 to allow the automated tests to access data in the new directory. 

6. The changes made required a few modifications to be compatible with HDF5. Included some notes in the documentation on how to use h5py to plot the HDF5 data. Currently, the back-transformed diagnostic in HDF5 format is not OpenPMD compliant. This will be done in a separate PR.

MaxThevenet commented:
Could you merge `dev` into this branch? I think that should fix the failing tests.

MaxThevenet commented on Source/Diagnostics/BoostedFrameDiagnostic.cpp:
```diff
@@ -1091,4 +1271,197 @@ writeSnapShotHeader() {
+    for (MFIter mfi(tmp, TilingIfNotGPU()); mfi.isValid(); ++mfi) {
+         Array4<Real> tmp_arr = tmp[mfi].array();
+         Array4<Real> buf_arr = buf[mfi].array();
+         // For 3D runs, tmp is a 2D (x,y) multifab that contains only
```
Could be 1D, 2D or 3D right?

MaxThevenet commented on Source/Diagnostics/BoostedFrameDiagnostic.H:
```diff
@@ -13,116 +13,231 @@
+    /// while for the reduced diagnostic, particles are selectively
+    /// copied if their trajectory crosses the user-defined
+    /// sub-domain for the reduced slice diagnostic.
+    virtual void AddPartDataToParticleBuffer(
```
Is the description above up-to-date? To my understanding, particles are copied if they were in the use-defined sub-domain. Is that correct? If so, would it be more accurate to say
```
    /// For the full diagnostic, all the particles are copied,
    /// while for the reduced diagnostic, particles are selectively
    /// copied if their are inside the user-defined
    /// sub-domain for the reduced slice diagnostic.
```
?

RevathiJambunathan replied:
Ah yes. This is correct. I have now changed it as follows :
    /// For the full diagnostic, all the particles are copied,
    /// while for the reduced diagnostic, particles are selectively
    /// copied if their position in within the user-defined
    /// sub-domain +/- 1 cell size width for the reduced slice diagnostic.

RevathiJambunathan commented on Source/Diagnostics/BoostedFrameDiagnostic.H:
```diff
@@ -13,116 +13,231 @@
+    /// sub-domain for the reduced slice diagnostic.
+    virtual void AddPartDataToParticleBuffer(
+                 amrex::Vector<WarpXParticleContainer::DiagnosticParticleData> tmp_particle_buffer,
+                 int nSpeciesBoostedFrame) {};
```
I have now added the override qualified for the functions defined in the derived class. Thanks for catching this!

RevathiJambunathan commented on Examples/Modules/RigidInjection/analysis_3Dbacktransformed_diag.py:
```diff
@@ -0,0 +1,37 @@
+Fs = allrd['Ez']
+Fs_1D = np.squeeze(Fs[Fs.shape[0]//2,Fs.shape[1]//2,:])
+
+error = np.max(np.abs(Fs_1D - F_1D))
```
Actually, Maxence had fixed this in his commit when fixing the automated tests. (I am responding to this comment just for completeness).

RevathiJambunathan commented on Examples/Modules/RigidInjection/analysis_3Dbacktransformed_diag.py:
```diff
@@ -0,0 +1,37 @@
+import numpy as np
+import scipy.constants as scc
+import read_raw_data
+yt.funcs.mylog.setLevel(0)
```
Maxence has fixed this in an earlier commit when fixing the automated tests. (Responding to this comment for completeness)

RevathiJambunathan commented on Docs/source/visualization/yt.rst:
```diff
@@ -124,6 +124,48 @@ Alternatively, the main commands can be found in our example jupyter
+    header   = './lab_frame_data/slices/Header'
+    allrd, info = read_raw_data.read_lab_snapshot(snapshot, header) # Read field data
+    F_RD = allrd[field]
+    plt.plot(F_RD[F_RD.shape[0]//2,F_RD.shape[1]//2,:])
```
This is addressed in the updated Docs.

RevathiJambunathan commented on Examples/Modules/RigidInjection/analysis_rigid_injection_BoostedFrame.py:
```diff
@@ -19,8 +19,8 @@
-snapshot = './lab_frame_data/snapshot00001'
-header   = './lab_frame_data/Header'
+snapshot = './lab_frame_data/snapshots/snapshot00001'
+header   = './lab_frame_data/snapshots/Header'
```
:)
